SteamKit2/SteamKit2/Steam/Handlers/SteamAuthTicket/SteamAuthTicket.cs

```diff
@@ -0,0 +1,208 @@
+            if ( Client.CellID == null ) throw new Exception( "User not logged in." );
+
+            var apps = Client.GetHandler<SteamApps>() ?? throw new Exception( "Steam Apps instance was null." );
+            var appTicket = await apps.GetAppOwnershipTicket( appid );
+
+            // User doesn't own the game
+            if ( appTicket.Result != EResult.OK ) throw new Exception( "User doesn't own the game." );
+
+            if ( GameConnectTokens.TryDequeue( out var token ) )
+            {
+                var authTicket = AuthTicketBuilder.Build( token );
+                var ticket = await VerifyTicket( appid, authTicket, out var crc );
+
+                // Verify just in case
+                if ( ticket.ActiveTicketsCRC.Any( x => x == crc ) )
+                {
+                    var tok = BuildTicket( authTicket, appTicket.Ticket );
+                    return new TicketInfo( this, appid, tok );
+                }
+                else
+                {
+                    throw new Exception( "Ticket verification failed." );
+                }
+            }
+            else
+            {
+                throw new Exception( "There's no available game connect tokens left." );
+            }
+        }
+
+        internal void CancelAuthTicket( TicketInfo authTicket )
+        {
+
+            if ( TicketsByGame.TryGetValue( authTicket.AppID, out var tickets ) )
+            {
+                tickets.RemoveAll( x => x.ticket_crc == authTicket.TicketCRC );
+            }
+
+            SendTickets();
+        }
+
+        private static byte[] BuildTicket( byte[] authTicket, byte[] appTicket )
+        {
+            var len = appTicket.Length;
+            var token = new byte[ authTicket.Length + 4 + len ];
+            var mem = token.AsSpan();
+            authTicket.CopyTo( mem );
+            MemoryMarshal.Write( mem[ authTicket.Length.. ], in len );
+            appTicket.CopyTo( mem[ ( authTicket.Length + 4 ).. ] );
+
+            return token;
+        }
+
+        private AsyncJob<TicketAcceptedCallback> VerifyTicket( uint appid, byte[] authToken, out uint crc )
+        {
+            crc = BitConverter.ToUInt32( Crc32.Hash( authToken ), 0 );
+            var items = TicketsByGame.GetOrAdd( appid, [] );
+
+            // Add ticket to specified games list
+            items.Add( new CMsgAuthTicket
+            {
+                gameid = appid,
+                ticket = authToken,
+                ticket_crc = crc
+            } );
+
+            return SendTickets();
+        }
+        private AsyncJob<TicketAcceptedCallback> SendTickets()
+        {
+            var auth = new ClientMsgProtobuf<CMsgClientAuthList>( EMsg.ClientAuthList );
+            auth.Body.tokens_left = ( uint )GameConnectTokens.Count;
+
+            auth.Body.app_ids.AddRange( TicketsByGame.Keys );
+            // Flatten dictionary into ticket list
+            auth.Body.tickets.AddRange( TicketsByGame.Values.SelectMany( x => x ) );
```

Comment:
getting keys and value separately doesn't sound safe considering it's a concurrent dictionary.